dotnet/src/Connectors/VectorData.Abstractions/ConnectorSupport/VectorStoreRecordVectorPropertyModel.cs

```diff
@@ -54,6 +60,147 @@ public int Dimensions
+        // (users who want arbitrary TInput types need to use the generic subclass of this type).
+        // The TEmbedding side is provided by the connector via the generic type parameter to this method, as the connector controls/knows which embedding types are supported.
+        // Note that if the user has manually specified an embedding type (e.g. to choose Embedding<Half> rather than the default Embedding<float>), that's provided via the embeddingType argument;
+        // we use that as a filter below.
+        switch (embeddingGenerator)
+        {
+            case IEmbeddingGenerator<string, TEmbedding> when this.Type == typeof(string) && (embeddingType is null || embeddingType == typeof(TUnwrappedEmbedding)):
+            case IEmbeddingGenerator<DataContent, TEmbedding> when this.Type == typeof(DataContent) && (embeddingType is null || embeddingType == typeof(TUnwrappedEmbedding)):
+                this.EmbeddingGenerator = embeddingGenerator;
+                this.EmbeddingType = embeddingType ?? typeof(TUnwrappedEmbedding);
+
+                return true;
+
+            case null:
+                throw new UnreachableException("This method should only be called when an embedding generator is configured.");
+            default:
+                return false;
+        }
+    }
+
+    /// <summary>
+    /// Attempts to generate an embedding of type <typeparamref name="TEmbedding"/> from the vector property represented by this instance on the given <paramref name="record"/>, using
+    /// the configured <see cref="EmbeddingGenerator"/>.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// If <see cref="EmbeddingGenerator"/> supports the given <typeparamref name="TEmbedding"/>, returns <see langword="true"/> and sets <paramref name="task"/> to a <see cref="Task"/>
+    /// representing the embedding generation operation. If <see cref="EmbeddingGenerator"/> does not support the given <typeparamref name="TEmbedding"/>, returns <see langword="false"/>.
+    /// </para>
+    /// <para>
+    /// The implementation on this non-generic <see cref="VectorStoreRecordVectorPropertyModel"/> checks for <see cref="string"/>
+    /// and <see cref="DataContent"/> as input types for <see cref="EmbeddingGenerator"/>.
+    /// </para>
+    /// </remarks>
+    public virtual bool TryGenerateEmbedding<TRecord, TEmbedding, TUnwrappedEmbedding>(TRecord record, CancellationToken cancellationToken, [NotNullWhen(true)] out Task<TEmbedding>? task)
+        where TRecord : notnull
+        where TEmbedding : Embedding
+    {
+        switch (this.EmbeddingGenerator)
+        {
+            case IEmbeddingGenerator<string, TEmbedding> generator when this.EmbeddingType == typeof(TUnwrappedEmbedding):
+            {
+                task = generator.GenerateEmbeddingAsync(
+                    this.GetValueAsObject(record) is var value && value is string s
```

Comment:
nit: since you are not using `value`, it  can be casted directly to string?

```suggestion
                    this.GetValueAsObject(record) is string s
```